### PR TITLE
Operator Publish full rewrite

### PR DESCRIPTION
**Pull request by akarnokd**
Complete rewrite of ```publish()``` and fixed bugs of #2810 and #2803.

### CODE REVIEW

**Comment by benjchristensen** [comment]
Wow. Will have to come back to this one :-) that's a pretty epic fix of my bad code!

**Comment by benjchristensen** [comment]
@davidmoten have you by chance tried this and been able to confirm it fixes your issues?

**Comment by akarnokd** [comment]
Ah, it was easy; you only need to dualize merge :).

**Comment by davidmoten** [comment]
@benjchristensen the original issue is covered by the rewrite, thanks @akarnokd . I've given the code a once over but not thorough, would take me a while and have little time at the moment.

**Comment by davidmoten** [comment]
@akarnokd I have had another brief foray into the code and I have every faith in your ability to create correct code but I'm not having a great time following it in the short slots of time I have between kid juggling. For longevity's sake I'm wondering if this beautiful looking bit of code could be bulked up with comments to lead me through and to lead future maintainers through. Of course if @benjchristensen or @zsxwing or another of the usual suspects think it's clear as day then further comments may be unnecessary.

**Comment by akarnokd** [comment]
Sure.

**Comment by akarnokd** [comment]
I've rebased, squashed and added comments to this PR.

**Comment by davidmoten** [comment]
In `OperatorPublishTest.testPublish` (old code, you made no changes), would be nice to have an assert between the calls to `subscribe` and the call to `connect` just to ensure that subscription has occured but nothing is emitted before the call to `connect`. This may be covered by other tests but not explicitly so I think could be worthwhile addition.

**Comment by akarnokd** [comment]
Thanks! Updated tests accordingly and fixed a case with clients starting with 0 requests.

**Comment by davidmoten** [comment]
Perhaps add a `connect` idempotency test?

**Comment by akarnokd** [comment]
Done.

File: src/main/java/rx/internal/operators/OperatorPublish.java [hunk 1]
```diff
@@ -15,367 +15,736 @@
...
+                        int len = ps.length;
+                        // Let's assume everyone requested the maximum value.
+                        long maxRequested = Long.MAX_VALUE;
+                        // count how many have triggered unsubscription
+                        int unsubscribed = 0;
+                    
+                        // Now find the minimum amount each child-subscriber requested
+                        // since we can only emit that much to all of them without violating
+                        // backpressure constraints
+                        for (InnerProducer<T> ip : ps) {
```
Comment (davidmoten) [inline src/main/java/rx/internal/operators/OperatorPublish.java]: This is one strategy I suppose and another might be request max but buffer responses for those subscribers that want less than max. I can imagine that either way somebody could suffer and I wonder if this strategy should be configurable?

**Comment by davidmoten** [comment]
By the way thanks for the comments in the code, they are just what I need!

File: src/main/java/rx/internal/operators/OperatorPublish.java (hunk 1 above)
Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorPublish.java]: You can only request the common minimum amount from upstream if the front-queue is bounded, even if some child subscribers request more. These may run ahead but then stay idle until the 'slower ones' catch up.

**Comment by akarnokd** [comment]
I'm preparing a JMH benchmark and it seems the current master fails with MissingBackpressureException on simple chains such as:
```java
@Test
public void testObserveOn() {
    ConnectableObservable<Integer> co = Observable.range(0, 1000).publish();
    Observable<Integer> obs = co.observeOn(Schedulers.computation());
    for (int i = 0; i < 100; i++) {
        TestSubscriber<Integer> ts = new TestSubscriber<Integer>();
        obs.subscribe(ts);
        Subscription s = co.connect();
        ts.awaitTerminalEvent(2, TimeUnit.SECONDS);
        ts.assertTerminalEvent();
        ts.assertNoErrors();
        assertEquals(1000, ts.getOnNextEvents().size());
        s.unsubscribe();
    }
}
```
I'll post and merge the PR for the perf test so it can be run independent of this PR.

**Comment by akarnokd** [comment]
I'll post its benchmark result but it takes a looong time. So far, I'm impressed (sync, 1 batch, 1 child, 1M elements):
master: 3.822 ops/s
this: 18.810 ops/s

**Comment by akarnokd** [comment]
[Here is](https://gist.github.com/akarnokd/8e4591bd4441cedb600e) the benchmark result for this PR in [JMH Comparison GUI](https://github.com/akarnokd/jmh-compare-gui) format. This PR outperforms the current master by 1.5 - 4x most of the time. However, some async parameter combination timed out despite the high CPU usage; I'll look into those cases.

File: src/main/java/rx/internal/operators/OperatorPublish.java [hunk 2]
```diff
@@ -15,367 +15,736 @@
...
+                     * Try adding it to the current subscriber-to-source, add is atomic in respect 
+                     * to other adds and the termination of the subscriber-to-source.
+                     */
+                    if (!r.add(inner)) {
+                        /*
+                         * The current PublishSubscriber has been terminated, try with a newer one.
+                         */
+                        continue;
+                        /*
+                         * Note: although technically corrent, concurrent disconnects can cause
```
Comment (benjchristensen) [inline src/main/java/rx/internal/operators/OperatorPublish.java]: By concurrent disconnect, I assume you mean concurrent subscribe/disconnect? If so then yes, the child subscribers may never receive anything if it is not re-connected. But that is exactly correct behavior if that's what happens. 

Am I understanding your comment here correctly?

File: src/main/java/rx/internal/operators/OperatorPublish.java [hunk 3]
```diff
@@ -15,367 +15,736 @@
...
+                                // meantime or hasn't requested anything yet
+                                // this eager behavior will skip unsubscribed children in case
+                                // multiple values are available in the queue
+                                if (ip.get() > 0L) {
+                                    try {
+                                        ip.child.onNext(value);
+                                    } catch (Throwable t) {
+                                        // we bounce back exceptions and kick out the child subscriber
+                                        ip.unsubscribe();
+                                        ip.child.onError(t);
```
Comment (benjchristensen) [inline src/main/java/rx/internal/operators/OperatorPublish.java]: I believe this is correct, but making sure my thinking is right. If the child `onError` itself throws (and breaks the contract) there is nothing we can do except blow up the entire thing, correct?

File: src/main/java/rx/internal/operators/OperatorPublish.java [hunk 4]
```diff
@@ -15,367 +15,736 @@
...
+                                    ip.produced(1);
+                                }
+                            }
+                            // indicate we emitted one element
+                            d++;
+                        }
+                        
+                        // if we did emit at least one element, request more to replenish the queue
+                        if (d > 0) {
+                            request(d);
```
Comment (benjchristensen) [inline src/main/java/rx/internal/operators/OperatorPublish.java]: Glad to see you requesting in batches.

File: src/main/java/rx/internal/operators/OperatorPublish.java [hunk 5]
```diff
@@ -15,367 +15,736 @@
...
+         */
+        static final long UNSUBSCRIBED = Long.MIN_VALUE;
+        /**
+         * Indicates this child has not yet requested any value. We pretend we don't
+         * see such child subscribers in dispatch() to allow other child subscribers who
+         * have requested to make progress. In a concurrent subscription scennario,
+         * one can't be sure when a subscription happens exactly so this virtual shift
+         * should not cause any problems.
+         */
+        static final long NOT_REQUESTED = Long.MIN_VALUE / 2;
```
Comment (benjchristensen) [inline src/main/java/rx/internal/operators/OperatorPublish.java]: This magic number made me smile :)

**Comment by benjchristensen** [comment]
Excellent code. Excellent comments. Very well done @akarnokd. This is a very hard piece of code. I can only imagine how much time you spent on this.

As best as I can tell it's all correct. It's complicated enough I may be missing something, but everything seems correct and matches what I wish I had coded this to be originally!

File: src/main/java/rx/internal/operators/OperatorPublish.java (hunk 2 above)
Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorPublish.java]: Yes, the behavior implemented and what you said is the correct one.

File: src/main/java/rx/internal/operators/OperatorPublish.java (hunk 3 above)
Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorPublish.java]: This is what happens, yes. I'm not sure if we want to go that deep and paranoid about onError and onCompleted.

**Pull request merged**